### PR TITLE
Moving all operator arguments to OperatorContext interface

### PR DESCRIPTION
**Pull request by frsyuki**
#340 made it possible to move `Operator.secretSelectors` to `OperatorFactory`.

This change takes advantage it and moves all arguments of operator to `OperatorContext` interface. Operator implementation will be:

```
class MyOperator
        extends BaseOperator
{
    MyOperator(OperatorContext context)
    {
        super(context);
    }

    @Override
    public TaskResult runTask()
    {
        ...
    }
}
```

A benefit is that we can add more fields to OperatorContext in the future without breaking backward compatibility of operator API.

This PR assumes #340. Diff from #340 is https://github.com/treasure-data/digdag/compare/privileged-vars...operator-context

### CODE REVIEW

File: digdag-core/src/main/java/io/digdag/core/agent/GrantedPrivilegedVariables.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+					Optional<Config> optionalNested = nested.getOptionalNested(nestName);
+                    if (!optionalNested.isPresent()) {
+                        if (required) {
+                            throw new ConfigException("Nested object '" + nestName + "' out of " + secretSharedKey + " is required but not set");
+                        }
+                        else {
+                            return null;
+                        }
+                    }
+					nested = optionalNested.get();
```
Comment (danielnorberg) [inline digdag-core/src/main/java/io/digdag/core/agent/GrantedPrivilegedVariables.java]: indentation

File: digdag-core/src/main/java/io/digdag/core/agent/GrantedPrivilegedVariables.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+            return (required) -> {
+                Optional<String> secret = secretProvider.getSecretOptional(secretOnlyKey.toString());
+                if (required && !secret.isPresent()) {
+                    throw new SecretNotFoundException(secretOnlyKey.toString());
+                }
+                return secret.orNull();
+            };
+        }
+        else {
+            ConfigKey secretSharedKey = grants.get(key, ConfigKey.class);
```
Comment (danielnorberg) [inline digdag-core/src/main/java/io/digdag/core/agent/GrantedPrivilegedVariables.java]: I'm a bit confused by the use of the word "shared". Can we come up with some name/wording that more accurately describes what's going on here?

File: digdag-core/src/main/java/io/digdag/core/agent/OperatorManager.java [hunk 3]
```diff
@@ -293,11 +294,19 @@ protected TaskResult callExecutor(Path projectPath, String type, TaskRequest mer
                 .build();
 
         DefaultSecretProvider secretProvider = new DefaultSecretProvider(
-                secretContext, secretAccessPolicy, grants, operatorSecretFilter, secretStore);
+                secretContext, secretAccessPolicy, grants, (key) -> declaredSecretKeys.contains(key), secretStore);
+
+        PrivilegedVariables privilegedVariables = GrantedPrivilegedVariables.build(
+                mergedRequest.getConfig().getNestedOrGetEmpty("_env"),
```
Comment (danielnorberg) [inline digdag-core/src/main/java/io/digdag/core/agent/OperatorManager.java]: Is `_env` a valid configuration key for all operators? It does not mean "environment" as in system environment variables?

File: digdag-standards/src/main/java/io/digdag/standards/operator/td/BaseTdJobOperator.java [hunk 4]
```diff
@@ -32,29 +33,30 @@
...
     }
 
-    @Override
-    public List<String> secretSelectors()
+    static ConfigSelector.Builder configSelectorBuilder()
     {
-        return ImmutableList.of("td.*");
+        return ConfigSelector.builderOfScope("td")
+            .addSecretOnlyAccess("apikey")
+            .addSecretSharedAccess("use_ssl", "proxy.enabled", "proxy.host", "proxy.port", "proxy.user", "proxy.password", "proxy.use_ssl", "endpoint", "host", "port", "user", "database");
```
Comment (danielnorberg) [inline digdag-standards/src/main/java/io/digdag/standards/operator/td/BaseTdJobOperator.java]: should `proxy.password` be secret only?

**Comment by danielnorberg** [comment]
👍  LTGM although I'd like to come up with a better name than "shared" for "secret _or_ param".

File: digdag-core/src/main/java/io/digdag/core/agent/DefaultOperatorContext.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    private final SecretProvider secretProvider;
+    private final PrivilegedVariables privilegedVariables;
+
+    DefaultOperatorContext(
+            Path projectPath,
+            TaskRequest taskRequest,
+            SecretProvider secretProvider,
+            PrivilegedVariables privilegedVariables)
+    {
+        this.projectPath = projectPath;
```
Comment (frsyuki) [inline digdag-core/src/main/java/io/digdag/core/agent/DefaultOperatorContext.java]: 👍

File: digdag-core/src/main/java/io/digdag/core/agent/GrantedPrivilegedVariables.java (hunk 1 above)
Comment (frsyuki) [inline digdag-core/src/main/java/io/digdag/core/agent/GrantedPrivilegedVariables.java]: 👍

File: digdag-core/src/main/java/io/digdag/core/agent/OperatorManager.java (hunk 3 above)
Comment (frsyuki) [inline digdag-core/src/main/java/io/digdag/core/agent/OperatorManager.java]: "_env" is set of environment variables given by workflow developers. Not by system administrators. commented more on https://github.com/treasure-data/digdag/pull/340#discussion_r82857853

File: digdag-core/src/main/java/io/digdag/core/agent/GrantedPrivilegedVariables.java (hunk 2 above)
Comment (frsyuki) [inline digdag-core/src/main/java/io/digdag/core/agent/GrantedPrivilegedVariables.java]: How about simply saying "secret access" and "secret-only access"? Public method names are:

* ConfigSelector.Builder.addSecretAccess(String...)
* ConfigSelector.Builder.addSecretOnlyAccess(String...)
* ConfigSelector.Builder.addAllSecretAccess(Iterable<? extends String>)
* ConfigSelector.Builder.addAllSecretOnlyAccess(Iterable<? extends String>)

Local variables can be like this:

```
+            ConfigKey secretAccessKey = grants.get(key, ConfigKey.class);
+            return (required) -> {
 +                Optional<String> secret = secretProvider.getSecretOptional(secretAccessKey.toString());
...
+                String value = nested.get(secretAccessKey.getLastName(), String.class, null);
```

**Comment by frsyuki** [comment]
rebased.
renamed `addSecretSharedAccess` to `addSecretAccess`.
pending merging until #340 is merged.

**Comment by danielnorberg** [comment]
:shipit:

**Comment by frsyuki** [comment]
Merged!

**Pull request merged**